lang/funcs/core/coresys/cpucount_fact.go

```diff
@@ -0,0 +1,201 @@
+func (obj CPUCountFact) Stream() error {
+	defer close(obj.init.Output) // Signal when we're done
+
+	ss, err := socketset.NewSocketSet(rtmGrps, socketFile, unix.NETLINK_KOBJECT_UEVENT)
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating socket set")
+	}
+
+	eventChan := make(chan *nlChanEvent) // updated when we receive uevent
+	defer close(eventChan)
+
+	// Start waiting for kernel to poke us about new
+	// device changes on the system
+	go func() {
+		defer ss.Shutdown()
+		defer ss.Close()
+		for {
+			uevent, err := ss.ReceiveUEvent()
+			//fmt.Printf("Sending %s\n", event.Data["SEQNUM"])
+			// TODO: use struct with error field
+			select {
+			case eventChan <- &nlChanEvent{
+				uevent: uevent,
+				err:    err,
+			}:
+			default: // don't block
+			}
+		}
+	}()
+
+	startChan := make(chan struct{})
+	close(startChan) // trigger the first event
+	var cpuCount int64
+	select {
+	case <-startChan:
+		startChan = nil // disable
+		cpuCount, _ = initCPUCount()
+		if err != nil {
+			// TODO: log?
+			cpuCount = 0
+		}
+		obj.init.Output <- &types.IntValue{
+			V: cpuCount,
+		}
+		break
+	case <-obj.closeChan:
+		return nil
+	}
+
+	for {
+		select {
+		case event, ok := <-eventChan:
+			if !ok {
+				continue
+			}
+			if event.err != nil {
+				// TODO: log here instead?
+				return errwrap.Wrapf(event.err, "error receiving uevent")
+			}
+			//fmt.Printf("Receiving %s\n", event.Data["SEQNUM"])
+			cpus, cpuEvent := processUdev(event.uevent)
+			if cpuEvent {
+				cpuCount += cpus
+				obj.init.Output <- &types.IntValue{
+					V: cpuCount,
+				}
+			}
+		case <-obj.closeChan:
+			return nil
+		}
+	}
+}
+
+// Info returns static typing info about what the fact returns
+func (obj *CPUCountFact) Info() *facts.Info {
+	return &facts.Info{
+		Output: types.NewType("int"),
+	}
+}
+
+// Close runs cleanup code for the fact and turns off the Stream.
+func (obj *CPUCountFact) Close() error {
+	close(obj.closeChan)
+	return nil
+}
+
+// initCPUCount looks in procfs to get the number of CPUs.
+// This is just for initializing the fact, and should not be polled.
+func initCPUCount() (int64, error) {
+	var count int64
+	dat, err := ioutil.ReadFile("/sys/devices/system/cpu/present") // TODO: change this to online?
+	if err != nil {
+		return 0, err
+	}
+
+	for _, line := range strings.Split(string(dat), ",") {
+		cpuRange := strings.SplitN(line, "-", 2)
+		if len(cpuRange) == 1 {
+			count++
+		} else if len(cpuRange) == 2 {
+			lo, err := strconv.ParseInt(cpuRange[0], 10, 64)
+			if err != nil {
+				return 0, err
+			}
+			hi, err := strconv.ParseInt(strings.TrimRight(cpuRange[1], "\n"), 10, 64)
+			if err != nil {
+				return 0, err
+			}
+			count += hi - lo + 1
+		}
+	}
+	return count, nil
+}
+
+// When we receieve a udev event, we filter only those that
+// indicate a CPU is being added or removed.
+// TODO: instead of checking for added/removed CPUS, check for online/offline?
```

Comment:
Yeah, so initially a cpu is added, and then it's made "online". That's when it becomes useful. So this is probably what we should use instead.

Alternatively, we could instead return a "big struct" of data about the cpus if you think that's useful. To be considered. Maybe better for a future patch.